bindings/matrix-sdk-ffi/src/room.rs

```diff
@@ -35,6 +35,52 @@ pub struct Room {
     timeline: RwLock<Option<Arc<Timeline>>>,
 }
 
+pub enum MembershipState {
+    /// The user is banned.
+    Ban,
+
+    /// The user has been invited.
+    Invite,
+
+    /// The user has joined.
+    Join,
+
+    /// The user has requested to join.
+    Knock,
+
+    /// The user has left.
+    Leave,
+}
+
+pub struct RoomMember {
```

Comment:
I think this can also exported using a uniffi macro, seems like `#[derive(uniffi::Record)]` or `#[derive(uniffi::Object)]` could be used.

@jplatte which one is more appropriate, doesn't seem that the derive macros have docs on them.

Reply:
Docs are [here](https://mozilla.github.io/uniffi-rs/proc_macro/index.html) now 🙂

`uniffi::Record` is the equivalent of `dictionary {}`, `uniffi::Object` is what you use in conjunction with `uniffi::export` on an impl as replacement for `interface {}` (although mixing `interface {}` with `uniffi::export` is also possible, see docs).

Reply:
Right, `Record` is probably then what we want here.

Reply:
I've tried it but couldn't make it build without the entry in udl file. I think it's because we use `RoomMember` on `Room` object in the udl.

Reply:
Build fails with:
```
thread 'main' panicked at 'Building the UDL file failed: Failed to parse UDL

Caused by:
    unknown type reference: RoomMember', bindings/matrix-sdk-ffi/build.rs:2:57
```

Reply:
Yes, if a type is referenced in UDL is must be defined in UDL. It is okay to do both though – use the derive so the type is usable in `#[uniffi::export]` fn's / impl's and define it in UDL so it is usable in fn signatures in the UDL.

Reply:
Updated, couldn't make it work with `Record` though so used `Object`.